### PR TITLE
[New Designs] Implement Dashboard Link update loading and Review Updates

### PR DESCRIPTION
**Pull request by aanorbel**
Fixes https://github.com/ooni/probe/issues/2594

## Proposed Changes

  - Add a worker to automatically update `descriptors` with `autoUpdate` enabled every 24 hrs.
  - Add a worker to fetch updates for `descriptors` with `autoUpdate` disabled and return to caller.
  - Show progress update for descriptors being fetched and prompt to review updates for descriptors with `autoUpdate` disabled in `MainActivity` 

|.|.|.|
|-|-|-|
| Screenshot_20240123_151744 | Screenshot_20240123_151752 | Screenshot_20240123_151808 |
| Screenshot_20240124_160104 | Screenshot_20240124_160101 | Screenshot_20240124_161236 |

### CODE REVIEW

**Comment by aanorbel** [comment]
I think we should make the interval for updates a little more configurable.

**Comment by hellais** [comment]
> I think we should make the interval for updates a little more configurable.

What is the interval currently?

Can we make them be updated every time the app is open and also before any background run?

**Comment by aanorbel** [comment]
Every time the app opens, I check updates for descriptors with Automated updates disabled and prompt the user for a review. For descriptors with auto-update enabled, I check for updates and update the descriptor every 24hrs.

**Comment by aanorbel** [comment]
> Can we make them be updated every time the app is opened and also before any background run?

Sure I can do that. It would be tricky to prompt the user to review updates if the updates were initiated before the background run starts since the new descriptors are kept in memory until the review is done. For auto updates, I don't see that being a problem.

File: app/src/main/java/org/openobservatory/ooniprobe/activity/AbstractActivity.java [hunk 1]
```diff
@@ -42,4 +46,18 @@ public void bindTestService() {
...
+     * This method is called when the task is completed.
+     */
+    protected void removeProgressFragment(@IdRes int id) {
+        Fragment fragment = getSupportFragmentManager().findFragmentById(id);
+        if (fragment != null && fragment.isAdded()) {
+            getSupportFragmentManager().beginTransaction().remove(fragment).commit();
+        }
+        findViewById(id).setVisibility(View.GONE);
+    }
+
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/AbstractActivity.java:62]:
```suggestion
```

File: app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java [hunk 2]
```diff
@@ -148,9 +162,105 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+                        AutoUpdateDescriptorsWorker.UPDATED_DESCRIPTORS_WORK_NAME,
+                        ExistingPeriodicWorkPolicy.KEEP,
+                        new PeriodicWorkRequest.Builder(AutoUpdateDescriptorsWorker.class, 24, TimeUnit.HOURS)
+                                .setConstraints(
+                                        new Constraints.Builder()
+                                                .setRequiredNetworkType(NetworkType.CONNECTED)
+                                                .build()
+                                ).build()
+                );
+        // TODO(aanorbel): add rules before checking updates
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java:181]: which rules?

File: app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java [hunk 3]
```diff
@@ -148,9 +162,105 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+                                    }
+
+                                    @Override
+                                    public void onCloseButtonClicked() {
+                                        removeProgressFragment(R.id.review_update_notification_fragment);
+                                    }
+                                }),
+                                OONIRunDynamicProgressBar.getTAG() + "_review_update_success_notification"
+                        ).commit();
+                case ENQUEUED -> getSupportFragmentManager()
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java:246]:
```suggestion

                case ENQUEUED -> getSupportFragmentManager()
```

File: app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java [hunk 4]
```diff
@@ -148,9 +162,105 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+                        .beginTransaction()
+                        .add(
+                                R.id.review_update_notification_fragment,
+                                OONIRunDynamicProgressBar.newInstance(ProgressType.UPDATE_LINK, null),
+                                OONIRunDynamicProgressBar.getTAG() + "_review_update_enqueued_notification"
+                        ).commit();
+                case FAILED -> Snackbar.make(
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java:253]:
```suggestion

                case FAILED -> Snackbar.make(
```

File: app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java [hunk 5]
```diff
@@ -148,9 +162,105 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+                        binding.getRoot(),
+                        R.string.Modal_Error,
+                        Snackbar.LENGTH_LONG
+                ).setAnchorView(binding.bottomNavigation).show();
+                default -> {
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java:258]:
```suggestion

                default -> {
```

I think we should document what is the default case with a comment. Is default === no updates?

File: app/src/main/java/org/openobservatory/ooniprobe/activity/OverviewActivity.java [hunk 6]
```diff
@@ -179,6 +194,72 @@ private void setUpOnCLickListeners() {
...
+                    binding.reviewUpdates.setVisibility(View.VISIBLE);
+                    binding.reviewUpdates.setOnClickListener(view -> getReviewUpdatesLauncher().launch(
+                            ReviewDescriptorUpdatesActivity.newIntent(
+                                    OverviewActivity.this,
+                                    workInfo.getOutputData().getString(ManualUpdateDescriptorsWorker.KEY_UPDATED_DESCRIPTORS)
+                            )
+                    ));
+                    binding.swipeRefresh.setRefreshing(false);
+                }
+                case FAILED -> {
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/OverviewActivity.java:251]:
```suggestion

                case FAILED -> {
```

File: app/src/main/java/org/openobservatory/ooniprobe/activity/OverviewActivity.java [hunk 7]
```diff
@@ -179,6 +194,72 @@ private void setUpOnCLickListeners() {
...
+                    binding.swipeRefresh.setRefreshing(false);
+                    Snackbar.make(
+                            binding.getRoot(),
+                            R.string.Modal_Error,
+                            Snackbar.LENGTH_LONG
+                    ).show();
+                }
+                default -> {
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/OverviewActivity.java:259]:
```suggestion

                default -> {
```

As commented above, we should explain what is the default case--no updates?

File: app/src/main/java/org/openobservatory/ooniprobe/activity/ReviewUpdatesAbstractActivity.kt [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+            registerForActivityResult(ActivityResultContracts.StartActivityForResult())
+            { result: ActivityResult ->
+                if (result.resultCode == RESULT_OK) {
+                    reviewCompletedListener()
+                    result.data?.let { intent: Intent ->
+                        intent.getStringExtra(ReviewDescriptorUpdatesActivity.RESULT_MESSAGE)
+                            ?.let { message: String ->
+                                Snackbar.make(view, message, Snackbar.LENGTH_LONG)
+                                    .setAnchorView(view).show()
+                            }
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/ReviewUpdatesAbstractActivity.kt:26]: Lisp vibes intensify!

File: app/src/main/java/org/openobservatory/ooniprobe/model/database/TestDescriptor.kt [hunk 9]
```diff
@@ -172,3 +184,61 @@ class NettestConverter : TypeConverter<String, Any>() {
     ).toList()
 }
 
+class ITestDescriptor(
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/model/database/TestDescriptor.kt:187]: What does the prefix `IT` mean in this context?

**Review by bassosimone — COMMENTED** [review]
This PR is large! 😅 

For now I only have questions and comments!

File: app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java (hunk 2 above)
Comment (aanorbel) [reply app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java:181]: Control for checking updates less frequently.

File: app/src/main/java/org/openobservatory/ooniprobe/model/database/TestDescriptor.kt (hunk 9 above)
Comment (aanorbel) [reply app/src/main/java/org/openobservatory/ooniprobe/model/database/TestDescriptor.kt:187]: The `I` is to denote its a type for interaction only. Not functional. It provides the proper types for all `TestDescriptor` properties.

Comment (bassosimone) [reply app/src/main/java/org/openobservatory/ooniprobe/model/database/TestDescriptor.kt:187]: gotcha!

File: app/src/main/java/org/openobservatory/ooniprobe/common/worker/UpdateDescriptorsWorker.kt [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+            for (descriptor in descriptors) {
+                Log.d(TAG, "Fetching updates for ${descriptor.runId}")
+
+                val updatedDescriptor: TestDescriptor =
+                    d.testDescriptorManager.fetchDescriptorFromRunId(
+                        descriptor.runId,
+                        applicationContext
+                    )
+                /**
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/common/worker/UpdateDescriptorsWorker.kt:121]: AFAIK, JavaDoc comments only have effect when documenting classes and methods. Did I got this incorrectly? Is this comment rendered in the documentation? If not, maybe I'd avoid using JavaDoc inside methods.

**Review by bassosimone — APPROVED** [review]
I have read this whole diff and it looks good to me

🐳

**Pull request merged**